### ISSUE TITLE
link error: undefined symbol of the EM_JS in EMSDK 3.1.28

### ISSUE THREAD

**Issue by xbcnn** [emscripten-core/emscripten#18436]
I have an EM_JS function defined in individual file and it works just fine before.  But recently I upgraded EMSDK to 3.1.28, I received linking error: undefined symbol of the EM_JS function.

I have opened EMCC_DEBUG=1, and got following errors.

`
profiler:DEBUG: block "JS symbol generation" took 0.391 seconds
wasm-ld: error: CMakeFiles/PRJ_dir/Unity/unity_5_cxx.cxx.o: undefined symbol: THE_EM_JS
`

I checked the `unity_5_cxx.cxx.o` file it does have the `__em_js_xxx` symbols.  Why the Binaryen failed to handle it?
How can I debug the Binaryen processing or maybe logs?

cc @sbc100 , any ideas?

**Comment by sbc100** [comment]
I think this is likely related to the fix for #18370.  Basically, after this change EM_JS functions appear like natively defined functions with a leading underscore.

I assume what is happening here is that you are using EM_JS functions from JS code or from object files other than the one in which they are defined... how are you declaring `THE_EM_JS` ?

Ultimately we might need to reverse that change to add the leading underscore to EM_JS functions if this effects many folks.

**Comment by sbc100** [comment]
@kleisauke was also effect my this issue (which I'm guessing is the extra leading underscore.

**Comment by kleisauke** [comment]
PR #18391 was landed as commit a74212999c52f6fa59b4f5452c3746087e6f7cf3 after Emscripten 3.1.28 was released. It's only available in the tip-of-tree build within emsdk at the time of writing.

Perhaps the issue is related to the `LLD_REPORT_UNDEFINED` change? FWIW, the Emscripten port of libffi removed the need of `-sLLD_REPORT_UNDEFINED=0` with commit https://github.com/hoodmane/libffi-emscripten/pull/20/commits/d1798d88fb955900b2452af3c28f487fa85b6808 (i.e. by implementing a native wrapper function).

**Comment by sbc100** [comment]
Ah yes, in that case this must be related instead to #16003.  

With this change all direct callers of EM_JS functions either need to exist in the same object file where it is defined, or they need to define their own `__attribute__(import_name('...'))` when they declare it.

**Comment by xbcnn** [comment]
> I assume what is happening here is that you are using EM_JS functions from JS code or from object files other than the one in which they are defined... how are you declaring `THE_EM_JS` ?

Yes. I use the EM_JS function in several other source files with extern "C" specifier and the exact name(no leading underscore) and parameters in the EM_JS.   Do you mean that it can be found with a leading underscore?

**Comment by sbc100** [comment]
#16003 did indeed lead to those symbols being mangled with a leading underscore.  However #18438 reverses that so that shouldn't be needed (assuming #18438 lands).

**Comment by xbcnn** [comment]
That's nice!  So I suppose it should be available in 3.1.29. #

**Comment by sbc100** [comment]
Yes, 3.1.29 should have the same behaviour as 3.1.28..  the change in questions was never part of a release.
